### PR TITLE
Static data endpoint

### PR DESCRIPTION
**Pull request by mathesond2**
This PR adds the following:

* all existing typechain types for hashes contracts (from `@hashes/types` dependency)
* ethers.js npm dependency
* `/util` dir with numerous util files for fns used now and in future
* top-level types file
* placeholder API route files for `/tokenData/dynamic` and `/walletData`
* `/tokenData/static` endpoint with tokenId as query param, endpoint response is as follows:
```json
{
  "hash": "0x5c2d581b129a5ef0cf91a04c2af0d1998e48074ea89485b327570242de06e4a5",
  "binary_value": "0101110000101101010110000001101100010010100110100101111011110000110011111001000110100000010011000010101011110000110100011001100110001110010010000000011101001110101010001001010010000101101100110010011101010111000000100100001011011110000001101110010010100101",
  "binary_attributes": [
    {
      "trait_type": "Leading 0's",
      "value": 1
    },
    {
      "trait_type": "Leading 1's",
      "value": 0
    },
    {
      "trait_type": "Trailing 0's",
      "value": 0
    },
    {
      "trait_type": "Trailing 1's",
      "value": 1
    },
    {
      "trait_type": "Longest Streak of 0's",
      "value": 8
    },
    {
      "trait_type": "Longest Streak of 1's",
      "value": 5
    },
    {
      "trait_type": "Max drift from center",
      "value": 29
    }
  ],
  "type": "DAO",
  "phrase_value": "satoshi",
  "phrase_attributes": [
    {
      "trait_type": "rarity",
      "value": 0.23112480739599386
    },
    {
      "trait_type": "word amount",
      "value": 1
    },
    {
      "trait_type": "character amount",
      "value": 7
    }
  ]
}
```

### CODE REVIEW

File: pages/api/tokenData/static.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  hash: string
+  binary_value: string
+  binary_attributes: BinaryAttribute[]
+  type: string
+  phrase_value: string
+  phrase_attributes: BinaryAttribute[]
+};
+
+function getHashesContract(chain_id: number | undefined): ethers.Contract {
+  const chainId = 1; //temp
```
Comment (mathesond2) [inline pages/api/tokenData/static.ts]: hitting mainnet for now

File: pages/api/tokenData/static.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  const { tokenId } = req.query;
+
+  if (isNaN(Number(tokenId))) {
+    res.status(400).send('tokenId must be a number');
+    return;
+  }
+
+  const hashesContract = getHashesContract(1);
+
+  const [hash, isDeactivated]: [string, boolean] = await Promise.all([
```
Comment (mathesond2) [inline pages/api/tokenData/static.ts]: How should we handle this potential err case?

File: pages/api/tokenData/static.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    res.status(404).send('token not found');
+    return;
+  }
+
+  const binaryValue = hex2bin(hash);
+  const binaryAttributes = getHashBinaryAttributes(hash);
+
+  //TODO: temp phrase solution for now
+  const generatedFilter = hashesContract.filters.Generated();
+  const AllGeneratedEvents = await hashesContract.queryFilter(generatedFilter);
```
Comment (mathesond2) [inline pages/api/tokenData/static.ts]: same thought here: How should we handle this potential err case?

File: util/hash_attributes.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import { getBinaryTreeData } from './binaryTreeBranch';
+
+export const HASH_ATTRIBUTES: Attribute[] = [
+    {
+        descriptionShort: "Leading 0's",
+        description: "Length of the streak of 0's at the start of the binary number representation.",
```
Comment (mathesond2) [inline util/hash_attributes.ts:6]: I left `description` in for now, we may use it?

**Review by mathesond2 — COMMENTED** [review]
This is a WIP PR...I still need to understand where we need to go re: any potential bannedlist per the perm3 spec.

I added a series of comments, would def. like some feedback re: the following:

- async err handling in `/static`endpoint
- any thoughts re: brevity of attribute responses, ex:
```js
{ type: val }
```
instead of:
```js
{
trait_type: "type",
value: val
}
```

File: util/hash_attributes.ts (hunk 4 above)
Comment (timsoulm) [reply util/hash_attributes.ts:6]: yeah definetly

**Comment by timsoulm** [comment]
> This is a WIP PR...I still need to understand where we need to go re: any potential bannedlist per the perm3 spec.
> 
> I added a series of comments, would def. like some feedback re: the following:
> 
> * async err handling in `/static`endpoint
> * any thoughts re: brevity of attribute responses, ex:
> 
> ```js
> { type: val }
> ```
> 
> instead of:
> 
> ```js
> {
> trait_type: "type",
> value: val
> }
> ```

The format is aligning to the opensea metadata standard so think it needs to be like this? https://docs.opensea.io/docs/metadata-standards#attributes

File: pages/api/tokenData/static.ts (hunk 2 above)
Comment (timsoulm) [reply pages/api/tokenData/static.ts]: eh probably just try/catch around the whole block for general unexpected errors

File: pages/api/tokenData/static.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  res: NextApiResponse<ResponseData | string>
+) {
+  const { tokenId } = req.query;
+
+  if (isNaN(Number(tokenId))) {
+    res.status(400).send('tokenId must be a number');
+    return;
+  }
+
+  const hashesContract = getHashesContract(1);
```
Comment (timsoulm) [inline pages/api/tokenData/static.ts:78]: you may want to get the 'nonce' from the hashes contract and check if the tokenId is greater than that. If it is, return early with a 404 or 400

File: pages/api/tokenData/static.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+  const phrase = tokenIdEvent?.args?.phrase;
+  const phraseAttributes = phrase ? getPhraseAttributes(phrase, AllGeneratedEvents) : [];
+
+  res.status(200).json({
+    hash,
+    binary_value: binaryValue,
+    binary_attributes: binaryAttributes,
+    type: Number(tokenId) >= 1000 ? 'Standard' : isDeactivated ? 'DAO Deactivated' : 'DAO',
+    phrase_value: phrase || 'Phrase is being non-fungibilitized... check back soon!',
```
Comment (timsoulm) [inline pages/api/tokenData/static.ts]: you wouldn't need this check if you're using the events. It only really matters if we're pulling from a DB which has a job where it gets written to every 15 mins (this covers the case where the token exists but the job hasnt run yet)

**Pull request merged**